### ISSUE TITLE
__name__ is unknown

### ISSUE THREAD

**Issue by Timmmm** [microsoft/pyright#2050]
Pylint gives errors when assigning to a class's `__name__`.

**To Reproduce**

Lint this code:

```python
class Foo:
    def __init__(self):
        self.__name__ = 'hello'
```

**Expected behavior**
No errors.

**Screenshots or Code**

![image](https://user-images.githubusercontent.com/376842/124289951-c723b100-db4a-11eb-900f-70a5ad72ef7a.png)

**VS Code extension or command-line**

Via Pylance v2021.6.3 using Python 3.8.

**Comment by erictraut** [comment]
This statement doesn't assign to the class's name. It's creating a new _instance_ variable called `__name__`.

The class variable `__name__` is unusual in Python. Unlike other class variables, it's not accessible through instances of that class. For example:

```python
print(int.__doc__) # OK
print(int.__name__) #OK

x = 1
print(x.__doc__) # OK to access this class variable from instance of `int`
print(x.__name__) # Runtime exception (AttributeError)
```

If you were to change your code sample above as follows, it will crash.
```python
class Foo:
    def __init__(self):
        print(self.__name__)

Foo()
```

Because of this odd behavior of `__name__`, pyright has special-case handling in place for this symbol. It generates an error when you try to access it from a class instance.

You're effectively overriding the class variable `__name__` with an instance variable of the same name, which apparently is allowed. I could add additional special-case logic to handle this case if there's a legitimate use case, but I'm not sure there is.

Why are you defining an instance variable called `__name__`? Since any attempt to access `__name__` from an instance is normally illegal and accessing it from the class (e.g. `Foo.__name__`) will access the underlying class variable, what are you trying to achieve here?

**Comment by Timmmm** [comment]
Thanks for the detailed reply! Sorry I am a Python neophyte. I didn't actually write this code - it's based [`LazyProperty` from here](http://blog.pythonisito.com/2008/08/lazy-descriptors.html).

**Comment by Timmmm** [comment]
If I change it to the following then it does work (Pyright no longer gives errors):

```
class Foo:
    __name__: str
    def __init__(self):
        self.__name__ = 'hello'
```

And I think it makes sense because there *is* an instance variable called `__name__`. Do you think that is a reasonable solution?

**Comment by erictraut** [comment]
I think that's a fine solution. Or even better, you could simply replace `__name__` with `_name`. It's just a private instance variable within the `LazyProperty` class, so I don't see any good reason to use the reserved (dundered) symbol `__name__` in this context. Dundered names (those that begin and end in double underscores) are reserved by the language and have special meaning.

**Comment by Timmmm** [comment]
Ah I was thinking that it had to set `__name__` so that the `__name__` of the property was passed though, but yeah it seems like you can never access it anyway, and I guess setting `__doc__` does nothing useful either. I guess it comes from this comment on the blog:

> also please copy the name to __name__ and the documentation to __doc__ so its more easy to use with the common introspection tools

Hmm. Thanks for the help anyway!
